### PR TITLE
AC-373 : Implement unsubscribtion in presenters

### PR DESCRIPTION
**Pull request by rhari991**
Added BasePresenter and ACBaseFragment that contain the logic related to subscribing/unsubscribing and setting the presenters. 

Refactored all the views and presents to extend these classes and removed repeated code.

### CODE REVIEW

**Comment by AdamGrzybkowski** [comment]
@rhari991 There are some lint errors. Could you fix them ?

**Comment by DefCon-007** [comment]
@rhari991  : It would be better if you squash all the unnecessary commits.

**Comment by AdamGrzybkowski** [comment]
@rhari991 Like @DefCon-007 said, you need to repair this PR. Not only squashing but there some commits that are not part of this PR. It probably happened when you were trying to resolve conflicts.

**Comment by tmarzeion** [comment]
PS. @rhari991 `git reflog` is useful tool when struggling with `git` :)
Just type `git reflog` and `git reset --hard <commit-hash>` to have instant time machine effect :)

**Comment by rhari991** [comment]
@tmarzeion thanks, any idea how I can see the lint errors that occured in travis ?

**Comment by tmarzeion** [comment]
@rhari991 run `./gradlew lint` within our project dir and see generated reports (with `*.html` extension)

**Comment by rhari991** [comment]
@tmarzeion thanks again :)
@AdamGrzybkowski Please review again

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/visitdashboard/VisitDashboardPresenter.java [hunk 1]
```diff
@@ -31,7 +32,7 @@
 import retrofit2.Response;
 import rx.android.schedulers.AndroidSchedulers;
 
-public class VisitDashboardPresenter implements VisitDashboardContract.Presenter {
+public class VisitDashboardPresenter extends BasePresenter implements VisitDashboardContract.Presenter {
```
Comment (AdamGrzybkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/visitdashboard/VisitDashboardPresenter.java]: what about endVisit() method in this presenter ?

Comment (rhari991) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/visitdashboard/VisitDashboardPresenter.java]: Looks like I missed that. I will add it in a minute.

Comment (rhari991) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/visitdashboard/VisitDashboardPresenter.java]: Done

**Comment by AdamGrzybkowski** [comment]
Looks good but could you rename the commit? There are some leftovers after resolving conflicts
`git rebase -i HEAD~1 -> chanage pick to r -> save file -> rename commit here -> save `

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/activevisits/ActiveVisitsFragment.java [hunk 2]
```diff
@@ -71,7 +69,6 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
     public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
         super.onViewCreated(view, savedInstanceState);
         mPresenter.updateVisitsInDatabaseList();
```
Comment (AdamGrzybkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/activevisits/ActiveVisitsFragment.java]: This is now called twice. In onResume() and here. I'm also not sure about doing subscription&unsubscription in those methods. E.g take a look here https://medium.com/@Viraj.Tank/art-of-subscribe-unsubscribe-in-rxjava-12c30d315777#.5yqce8q35

Comment (rhari991) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/activevisits/ActiveVisitsFragment.java]: Shall I move the calls to onViewCreated() and onDestroyView() ?

Comment (AdamGrzybkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/activevisits/ActiveVisitsFragment.java]: yeah I think we should go this way

Comment (rhari991) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/activevisits/ActiveVisitsFragment.java]: Done

**Comment by AdamGrzybkowski** [comment]
@rhari991 please fix commit message again :) Good command to add some small changes is `git commit --am` it will add changes to the last commit

**Comment by rhari991** [comment]
What should I change the commit message to ?

**Comment by AdamGrzybkowski** [comment]
@rhari991 just delete "small change" form the end of the message

**Comment by rhari991** [comment]
Sorry about that, I am pretty new to using Git

**Comment by AdamGrzybkowski** [comment]
@rhari991 No problem at all :) I'm not an expert either and I've been there too - struggling with rebasing :dagger:

**Comment by rhari991** [comment]
Is it good now ?

**Comment by AdamGrzybkowski** [comment]
@rhari991 I did some manual testing and we introduced a bug by calling subscribe() in onviewCreated(). You can reproduce it by going to patient dashboard, starting new visit and then pressing back button. You won't see updated list and that means onviewCreated is not called which means we have to revert it to using onResume/onPause.

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/formdisplay/FormDisplayPageFragment.java [hunk 3]
```diff
@@ -63,8 +62,7 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
                 WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN);
 
         mParent = (LinearLayout) root.findViewById(R.id.sectionContainer);
-        mPresenter.start();
-
+        mPresenter.subscribe();
```
Comment (AdamGrzybkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/formdisplay/FormDisplayPageFragment.java]: This is duplicated

Comment (rhari991) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/formdisplay/FormDisplayPageFragment.java]: removed

**Pull request merged**